server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAllocationHealthIndicatorService.java

```diff
@@ -0,0 +1,164 @@
+    }
+
+    private static class ShardAllocationCounts {
+        private int unassigned = 0;
+        private int unassigned_restarting = 0;
+        private int initializing = 0;
+        private int started = 0;
+        private int relocating = 0;
+
+        public void increment(ShardRouting routing) {
+            switch (routing.state()) {
+                case UNASSIGNED -> {
+                    unassigned++;
+                    if (routing.unassignedInfo() != null && routing.unassignedInfo().getReason() == UnassignedInfo.Reason.NODE_RESTARTING) {
+                        unassigned_restarting++;
+                    }
+                }
+                case INITIALIZING -> initializing++;
+                case STARTED -> started++;
+                case RELOCATING -> relocating++;
+            }
+        }
+    }
+
+    private static class ShardAllocationStatus {
+        private final ShardAllocationCounts primaries = new ShardAllocationCounts();
+        private final ShardAllocationCounts replicas = new ShardAllocationCounts();
+
+        public void addPrimary(ShardRouting routing) {
+            primaries.increment(routing);
+        }
+
+        public void addReplica(ShardRouting routing) {
+            replicas.increment(routing);
+        }
+
+        public HealthStatus getStatus() {
+            if (primaries.unassigned > 0) {
+                return RED;
+            } else if (replicas.unassigned > replicas.unassigned_restarting) {
+                return YELLOW;
+            } else {
+                return GREEN;
+            }
+        }
+
+        public String getSummary() {
+            var builder = new StringBuilder("This cluster has ");
+            if (primaries.unassigned > 0 || replicas.unassigned > 0) {
+
+                builder.append(
+                    Stream.of(
+                        Stream.of("1 unassigned primary").filter(it -> primaries.unassigned == 1),
+                        Stream.of(primaries.unassigned + " unassigned primaries").filter(it -> primaries.unassigned > 1),
+                        Stream.of("1 unassigned replica").filter(it -> replicas.unassigned == 1),
+                        Stream.of(replicas.unassigned + " unassigned replicas").filter(it -> replicas.unassigned > 1)
+                    ).flatMap(Function.identity()).collect(joining(" and "))
+                ).append(".");
+            } else {
+                builder.append("no unassigned shards.");
+            }
+            return builder.toString();
+        }
+
+        public SimpleHealthIndicatorDetails getDetails() {
+            return new SimpleHealthIndicatorDetails(
+                Map.of(
+                    "unassigned_primaries",
+                    primaries.unassigned,
+                    "initializing_primaries",
+                    primaries.initializing,
+                    "started_primaries",
+                    primaries.started,
+                    "relocating_primaries",
```

Comment:
I guess I mostly don't know if relocating shards matter for this PR? Should we be including diagnostic information that does not pertain to shards being unavailable?

Reply:
I guess this is not strictly related, though if cloud is going to use this indicator to populate shard status pie chart, then they would also need reallocating shards